File: app/src/main/java/com/example/finanzas/data/api/CategoryBudgetService.java
package com.example.finanzas.data.api;

import android.content.Context;

import com.example.finanzas.data.local.LocalRepository;
import com.example.finanzas.data.model.CategoryBudgetSummary;

import java.util.List;

public class CategoryBudgetService {

    public interface ListCb { void onOk(List<CategoryBudgetSummary> items); void onFail(); }
    public interface SaveCb { void onOk(); void onFail(); }

    public static void list(Context ctx, int anio, int mes, ListCb cb) {
        try {
            List<CategoryBudgetSummary> list = LocalRepository.getInstance(ctx).listPresupuestosCategoria(anio, mes);
            cb.onOk(list);
        } catch (Exception e) {
            cb.onFail();
        }
    }

    public static void save(Context ctx, int anio, int mes, List<?> items, SaveCb cb) {
        try {
            LocalRepository.getInstance(ctx).savePresupuestosCategoria(anio, mes, items);
            cb.onOk();
        } catch (Exception e) {
            cb.onFail();
        }
    }
}

File: app/src/main/java/com/example/finanzas/data/api/CategoryBudgetService.kt
package com.example.finanzas.data.api

import android.content.Context
import com.example.finanzas.data.local.LocalRepository
import com.example.finanzas.data.model.CategoryBudgetInput
import com.example.finanzas.data.model.CategoryBudgetSummary
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

object CategoryBudgetService {

    interface ListCb {
        fun onOk(items: List<CategoryBudgetSummary>)
        fun onFail()
    }

    interface SaveCb {
        fun onOk()
        fun onFail()
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    suspend fun list(ctx: Context, anio: Int, mes: Int): List<CategoryBudgetSummary> = withContext(Dispatchers.IO) {
        LocalRepository.getInstance(ctx).listPresupuestosCategoria(anio, mes)
    }

    suspend fun save(ctx: Context, anio: Int, mes: Int, items: List<*>) = withContext(Dispatchers.IO) {
        LocalRepository.getInstance(ctx).savePresupuestosCategoria(anio, mes, items)
    }

    suspend fun savedInputs(ctx: Context, anio: Int, mes: Int): List<CategoryBudgetInput> = withContext(Dispatchers.IO) {
        LocalRepository.getInstance(ctx).listSavedPresupuestosCategoriaInputs(anio, mes)
    }

    @JvmStatic
    fun list(ctx: Context, anio: Int, mes: Int, cb: ListCb) {
        scope.launch {
            runCatching { list(ctx, anio, mes) }
                .onSuccess(cb::onOk)
                .onFailure { cb.onFail() }
        }
    }

    @JvmStatic
    fun save(ctx: Context, anio: Int, mes: Int, items: List<*>, cb: SaveCb) {
        scope.launch {
            runCatching { save(ctx, anio, mes, items) }
                .onSuccess { cb.onOk() }
                .onFailure { cb.onFail() }
        }
    }
}
